### PR TITLE
Don't check trivial dirty bits

### PR DESCRIPTION
**Pull request by pavpanchekha**
This PR changes Chapter 16 to only check dirty bits for:

- Rebuilding the layout tree
- Skipping `layout` entirely
- Recomputing `style`

The text changes corresponding to this are probably not the best I could do—I plan to leave that for another PR that does a full read of the chapter.

### CODE REVIEW

File: book/invalidation.md [hunk 1]
```diff
@@ -446,7 +446,7 @@ array isn't dirty. Let's find the line at the top of `layout` that
 resets the `children` array, and move it into the two branches of the
 `if` statement:
 
-``` {.python expected=False}
+``` {.python}
```
Comment (pavpanchekha) [inline book/invalidation.md:449]: I dunno why this changed but I don't want to mess with it at the moment.

Comment (pavpanchekha) [reply book/invalidation.md:449]: (As in, I don't know why this is now matching, but I don't care.)

File: book/invalidation.md [hunk 2]
```diff
@@ -947,12 +945,17 @@ class BlockLayout:
     def layout(self):
         # ...
         if mode == "block":
-            # ...
+            if self.children.dirty:
+                # ...
```
Comment (pavpanchekha) [inline book/invalidation.md:949]: Also a random change to diffing that shouldn't be affected by this chapter but somehow is.

File: src/lab16.py [hunk 3]
```diff
@@ -233,26 +233,34 @@ def __init__(self, node, frame):
...
+        if self.width.dirty: return True
+        if self.height.dirty: return True
+        if self.x.dirty: return True
+        if self.y.dirty: return True
+        if self.descendants.dirty: return True
+        return False
+
     def layout(self, width, zoom):
         self.zoom.set(zoom)
+        self.width.set(width - 2 * device_px(HSTEP, zoom))
```
Comment (chrishtr) [inline src/lab16.py:247]: Why put these two lines before the layout_needed check?

Comment (pavpanchekha) [reply src/lab16.py:247]: Well, if the `zoom` or `width` argument has changed, we need to update those fields. The `layout_needed` method will check their dirty bits, but not diff the current value with the argument. We could change how zooming / resizing works but I think this is easier.

Comment (chrishtr) [reply src/lab16.py:247]: If `layout_needed` checks `self.zoom.dirty` and `self.width.dirty`, then i'll always recompute layout if they change, right?

Comment (pavpanchekha) [reply src/lab16.py:247]: I think I figured out the confusion. `zoom` and `width` aren't modified or marked by the `Frame` when the zoom level or frame width changes. (So they're not dirty until the `set` operation.) Would you rather do that—mark those fields dirty when the inputs change?

Comment (chrishtr) [reply src/lab16.py:247]: That would be more consistent, so let's try that.

Comment (pavpanchekha) [reply src/lab16.py:249]: Done. Honestly, I think this approach is a little worse, but take a look and let me know. It's not a huge difference either way.

File: book/invalidation.md [hunk 4]
```diff
@@ -1926,24 +1926,39 @@ class ProtectedField:
...
-    def layout(self, width, zoom):
-        self.zoom.set(zoom)
-        self.width.set(width - 2 * device_px(HSTEP, zoom))
-        if not self.layout_needed(): return
+class IframeLayout(EmbedLayout):
+    def layout(self):
+        if self.node.frame:
+            # ...
+            self.node.frame.document.height.mark()
+            self.node.frame.set_needs_render()
```
Comment (chrishtr) [inline book/invalidation.md:1990]: why is set_needs_render needed here?

Comment (pavpanchekha) [reply book/invalidation.md:1990]: So, it at least needs to `set_needs_layout`, right?

**Comment by pavpanchekha** [comment]
Decided: remove the `set_needs_render` line but add a sentence to Chapter 15 about why we get to skip it.

**Comment by pavpanchekha** [comment]
Changes decided on in meeting have been done: no more `set_needs_render`, and two footnotes about:

1. Why not call it.
2. Why not protect `frame_width` instead.

File: book/embeds.md [hunk 5]
```diff
@@ -954,7 +954,14 @@ class Frame:
         self.frame_height = 0
 ```
 
-And we can set those when the parent frame is laid out:
+And we can set those when the parent frame is laid out:[^no-set-needs]
+
+[^no-set-needs]: You might be surprised that I'm not calling
```
Comment (chrishtr) [inline book/embeds.md:959]: I just remembered that chapter 14 introduces `setAttribute`. So this is not true?

Comment (chrishtr) [reply book/embeds.md:959]: I just put up a PR to call set_needs_render from setAttribute. So just rephrase this footnote
to refer to the call there?

**Comment by pavpanchekha** [comment]
Ok, I fixed the footnote. Let's test and, if necessary, fix Chapter 16 in another PR.

**Pull request merged**